### PR TITLE
feat(Hubspot Node): enable hubspot credentials for http predefined types

### PR DESCRIPTION
**Pull request by vcrwr**
Changed Hubspot ApiKey and AppToken credentials to be supported as httpRequest Node "Predefined Credential Type".

### CODE REVIEW

**Comment by vcrwr** [comment]
With the deprecation of Hubspot APIkeys, AppToken are getting important very soon. Would be great if you could review and accept the PR. :)

**Comment by Joffcom** [comment]
Hey @vcrwr,

I am not sure how I missed this one but I will get this one sorted later today 👍🏻

**Comment by Joffcom** [comment]
Hey @vcrwr,

There are a couple of changes I want to make but it won't let me add to the PR, I am not sure why though. If I pop them into the PR as nodes do you fancy doing them unless you have an option to enable write access on the PR.

**Comment by vcrwr** [comment]
According to the documentation, there should be a "Allow edits from maintainers" option in the PR for me. But there is not. :(

Of course I'm happy to apply your changes.

File: packages/nodes-base/credentials/HubspotApi.credentials.ts [hunk 1]
```diff
@@ -12,4 +12,13 @@ export class HubspotApi implements ICredentialType {
...
+
+	authenticate: IAuthenticateGeneric = {
+		type: 'generic',
+		properties: {
+			qs: {
+				hapikey: '={{$credentials.apiKey}}',
+			},
+		},
+	};
 }
```
Comment (Joffcom) [inline packages/nodes-base/credentials/HubspotApi.credentials.ts:24]: Add Credential Test

```
test: ICredentialTestRequest = {
		request: {
			baseURL: 'https://api.hubapi.com',
			url: '/account-info/v3/details',
		},
	};
```

File: packages/nodes-base/credentials/HubspotAppToken.credentials.ts [hunk 2]
```diff
@@ -12,4 +12,13 @@ export class HubspotAppToken implements ICredentialType {
...
+
+	authenticate: IAuthenticateGeneric = {
+		type: 'generic',
+		properties: {
+			headers: {
+				Authorization: '=Bearer {{$credentials.appToken}}',
+			},
+		},
+	};
 }
```
Comment (Joffcom) [inline packages/nodes-base/credentials/HubspotAppToken.credentials.ts:24]: Add Credential Test

```
	test: ICredentialTestRequest = {
		request: {
			baseURL: 'https://api.hubapi.com',
			url: '/account-info/v3/details',
		},
	};
```

File: packages/nodes-base/nodes/Hubspot/GenericFunctions.ts [hunk 3]
```diff
@@ -39,15 +39,9 @@ export async function hubspotApiRequest(this: IHookFunctions | IExecuteFunctions
 
 	try {
 		if (authenticationMethod === 'apiKey') {
```
Comment (Joffcom) [inline packages/nodes-base/nodes/Hubspot/GenericFunctions.ts:41]: We can make this a bit smaller and just have...
```
if (authenticationMethod === 'apiKey' || authenticationMethod === 'appToken') {
			return this.helpers.requestWithAuthentication.call(this, authenticationMethod, options);
		} else if (authenticationMethod === 'developerApi') {
```

**Review by Joffcom — CHANGES_REQUESTED** [review]
Notes below, Thanks again for this one.

**Comment by vcrwr** [comment]
@Joffcom thanks for the review. I added the test requests and changed the code as recommended.

**Comment by Joffcom** [comment]
Hey @vcrwr,

Looks good, I will give it another quick test shortly just to make sure it all works then we can get it merged in 👍🏻

File: packages/nodes-base/nodes/Hubspot/GenericFunctions.ts [hunk 4]
```diff
@@ -38,16 +38,8 @@ export async function hubspotApiRequest(this: IHookFunctions | IExecuteFunctions
...
-
-			options.qs.hapikey = credentials.apiKey as string;
-			return await this.helpers.request!(options);
-		} else if (authenticationMethod === 'appToken') {
-			const credentials = await this.getCredentials('hubspotAppToken');
-
-			options.headers!['Authorization'] = `Bearer ${credentials.appToken}`;
-			return await this.helpers.request!(options);
+		if (authenticationMethod === 'apiKey' || authenticationMethod === 'appToken') {
+			return this.helpers.requestWithAuthentication.call(this, authenticationMethod, options);
```
Comment (Joffcom) [inline packages/nodes-base/nodes/Hubspot/GenericFunctions.ts:42]: Hey @vcrwr,

Just testing and I realised I messed up... This needs to be...

```
const credentialType = authenticationMethod === 'apiKey' ? 'hubspotApi' : 'hubspotAppToken';
return this.helpers.requestWithAuthentication.call(this, credentialType, options);
```

I forgot that the Hubspot node is one of the few where the authentication method doesn't match the actual credential name.

If you can get this part added in the rest of it is good and I can get it merged.

Comment (vcrwr) [reply packages/nodes-base/nodes/Hubspot/GenericFunctions.ts:42]: Thats weird. I saw this and then decided to test it manually too. I was able to pull data from the API with both credentialTypes. Let me check, what I did wrong on the testing side.

Comment (vcrwr) [reply packages/nodes-base/nodes/Hubspot/GenericFunctions.ts:42]: In the moment I wrote it, I see the issue. I tested with the HttpNode. ;)

Comment (vcrwr) [reply packages/nodes-base/nodes/Hubspot/GenericFunctions.ts:42]: While thinking about it, I'd rather would switch back to the original variant. The need for another variable is symbolic for the increase of complexity. Its less readable in the new version IMHO. Of course the best way would be, to adjust the Hubspot Credential Naming, but as this breaks with existing installations, this is surely not an option.

@Joffcom what do you think about rolling back to the original if else if variant?

Comment (Joffcom) [reply packages/nodes-base/nodes/Hubspot/GenericFunctions.ts:42]: @vcrwr I like the smaller version as it reduces reused code. But both ways would work.

Comment (vcrwr) [reply packages/nodes-base/nodes/Hubspot/GenericFunctions.ts:42]: Ok. Happy to apply your smaller version then. I'll give it a bit more testing tomorrow, before committing.

Comment (Joffcom) [reply packages/nodes-base/nodes/Hubspot/GenericFunctions.ts:42]: I will give it another test as well just to be safe. Thanks for going through this.

Comment (vcrwr) [reply packages/nodes-base/nodes/Hubspot/GenericFunctions.ts]: smaller version committed and tested successfully.

**Review by Joffcom — APPROVED** [review]

**Comment by Joffcom** [comment]
Thanks for the PR @vcrwr this is now merged 🙌🏻

**Comment by janober** [comment]
Got released with `n8n@0.199.0`

**Pull request merged**